File: Blind 75/121. Best Time to Buy and Sell Stock.ts
function maxProfit(prices: number[]): number {
  let buyPrice = prices[0];
  let highestProfit = 0;

  for (let price of prices) {
    if (price < buyPrice) {
      buyPrice = price;
    } else {
      highestProfit = Math.max(price - buyPrice, highestProfit);
    }
  }
  return highestProfit;
}

File: Blind Golang/Array/121. Best Time to Buy and Sell Stock.go
package array

func maxProfit(prices []int) int {
	var highestProfit int
	buyPrice := &prices[0]

	for i := 1; i < len(prices); i++ {
		if diff := prices[i] - *buyPrice; diff > highestProfit {
			highestProfit = diff
		}

		if prices[i] < *buyPrice {
			buyPrice = &prices[i]
		}
	}

	return highestProfit
}
